duplex_http_call.go

```diff
@@ -316,3 +339,77 @@ func cloneURL(oldURL *url.URL) *url.URL {
 	}
 	return newURL
 }
+
+// bufferPipeWriter is a write-through cache.
+type bufferPipeWriter struct {
+	*io.PipeWriter
+
+	mu     sync.Mutex
+	buffer *bytes.Buffer
+}
+
+// Write to the PipeWriter and buffer the data if available.
+func (b *bufferPipeWriter) Write(data []byte) (int, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	bytesWritten, err := b.PipeWriter.Write(data)
+	if b.buffer == nil || err != nil {
+		return bytesWritten, err
+	}
+	_, _ = b.buffer.Write(data[:bytesWritten])
+	if b.buffer.Len() > maxRecycleBufferSize {
+		b.buffer = nil // Don't recycle large buffers.
```

Comment:
Wanted to limit the buffersizes and potential perf impact, maybe not needed.

Reply:
I think you may have missed my point. The issue with this is that we _never_ can successfully retry an operation where a request is >8mb. I think that's a bad side effect here that suggests an alternate approach may be needed.

Reply:
We have to have a limit somewhere, it could be be based on the message size limit so first message would always succeed. How would we handle client streams that always read?

Reply:
> always read but never write
Sorry, always write but never read. If the writer is responsible for releasing the memory what do we use as the upper limit?